java/arcs/core/storage/StorageProxy.kt

```diff
@@ -387,16 +396,20 @@ class StorageProxy<Data : CrdtData, Op : CrdtOperationAtTime, T>(
         deferred: CompletableDeferred<Boolean> = CompletableDeferred()
     ) {
         val queueNum = outgoingMessagesInFlight.getAndIncrement()
-        log.debug { "Queueing message (pos: $queueNum) for sending to the store: $message" }
         if (queueNum == 0) {
             busySendingMessagesChannel.offer(true)
         }
-        outgoingMessagesChannel.offer(message to deferred)
+        do {
+            val sent = outgoingMessagesChannel.offer(message to deferred)
+            log.debug {
+                "Queueing ${if (sent) "" else "un"}successful for message (pos: $queueNum) " +
+                    "for sending to the store: $message"
+            }
+        } while (!sent)
```

Comment:
What are the failure modes like here? Is there a chance this will just start spewing a log message endlessly?

Reply:
Looks like the internals of `send` (the version which suspends until it can enqueue the item) also does a loop.  https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/common/src/channels/AbstractChannel.kt#L168-L202

I could drop the log until after the loop exits though.